pkg/controller/direct/directbase/directbase_controller.go

```diff
@@ -53,22 +54,41 @@ import (
 	"sigs.k8s.io/controller-runtime/pkg/source"
 )
 
-// Add creates a new controller for reconciling objects of the specified GVK, delegating actual resource reconciliation to the provided Model.
-func Add(mgr manager.Manager, gvk schema.GroupVersionKind, model Model, opts Deps) error {
+var ControllerBuilder directControllerBuilder
+
+func init() {
+	ControllerBuilder = directControllerBuilder{modelMapper: map[schema.GroupVersionKind]func(*controller.Config) Model{}}
+}
+
+type directControllerBuilder struct {
+	modelMapper map[schema.GroupVersionKind]func(*controller.Config) Model
+}
+
+func (c *directControllerBuilder) RegisterModel(gvk schema.GroupVersionKind, modelFn func(*controller.Config) Model) {
+	c.modelMapper[gvk] = modelFn
```

Comment:
Nit: One trick, if it isn't perf sensitive (and this isn't, I believe), is to check if modelMapper is nil and initialize the map here.

Reply:
Sg. Thanks!